src/dune_engine/load_rules.ml

```diff
@@ -524,25 +535,89 @@ end = struct
         ~subdir:(Path.Build.basename dir)
   end
 
-  let load_dir_step2_exn ~dir ~context_or_install ~sub_dir =
-    let sub_dir_components = Path.Source.explode sub_dir in
+  module rec Gen_rules : sig
+    val gen_rules :
+      Dir_triage.Build_directory.t -> (Subdir_set.t * Rules.t) Memo.Build.t
+  end = struct
+    let gen_rules_impl
+        { Dir_triage.Build_directory.dir; context_or_install; sub_dir } =
+      let (module RG : Build_config.Rule_generator) =
+        (Build_config.get ()).rule_generator
+      in
+      let sub_dir_components = Path.Source.explode sub_dir in
+      RG.gen_rules context_or_install ~dir sub_dir_components >>= function
+      | Rules (subdirs, rules) -> (
+        match
+          Path.Build.Map.find_key (Rules.to_map rules) ~f:(fun p ->
+              not (Path.Build.is_descendant p ~of_:dir))
+        with
+        | None -> Memo.Build.return (subdirs, rules)
+        | Some p ->
+          let dir_rules =
+            Rules.find rules (Path.build p) |> Rules.Dir_rules.consume
+          in
+          Code_error.raise
+            "[gen_rules] returned rules in a directory that is not a \
+             descendant of the directory it was called for"
+            [ ("dir", Path.Build.to_dyn dir)
+            ; ( "example"
+              , match dir_rules with
+                | { rules = r :: _; _ } ->
+                  Dyn.Variant
+                    ( "Rule"
+                    , [ Dyn.Record
+                          [ ("targets", Targets.Validated.to_dyn r.targets) ]
+                      ] )
+                | { rules = []; aliases } -> (
+                  match Alias.Name.Map.choose aliases with
+                  | None -> assert false
+                  | Some (name, _) ->
+                    Dyn.Variant
+                      ( "Alias"
+                      , [ Dyn.Record
+                            [ ("dir", Path.Build.to_dyn p)
+                            ; ("name", Alias.Name.to_dyn name)
+                            ]
+                        ] )) )
+            ])
+      | Unknown_context_or_install ->
```

Comment:
Does it make sense to push this `Code_error` into `gen_rules` to avoid the extra constructor and have the error and the relevant logic closer to each other?

Reply:
Alternatively, I guess it makes sense for `gen_rules` to have `context` as part of the input `Build_directory`. It seems to me that it's the job of `get_dir_triage` to validate the context name.

Reply:
Oh, I guess `get_dir_triage` doesn't have enough information to do so...

Reply:
Agreed. I'll leave that for another PR.